### PR TITLE
Add mulit-variable optimization example  

### LINKED ISSUE #27807: Optimize mechanical load with both displacment measurement points
### Motivation

Show the new, simplified way for optimizing a problem with multiple measurement points of different variables.

### Design

Setup example input file.

### Impact

Drastically simplifies the input file needed for multi-variable optimization problems.

### PR DESCRIPTION
**Pull request by maxnezdyur**
closes #27807

### CODE REVIEW

**Comment by maxnezdyur** [comment]
@lynnmunday 
here is the example we talked about.

File: modules/combined/test/tests/optimization/invOpt_mechanics/forward_and_adjoint.i [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+  [disp_x]
+  []
+  [disp_y]
+  []
+[]
+
+[Physics]
+  [SolidMechanics]
+    [QuasiStatic]
+      displacements = 'disp_x disp_y'
```
Comment (lynnmunday) [inline modules/combined/test/tests/optimization/invOpt_mechanics/forward_and_adjoint.i:23]: Why do you need this. I have to do this too?  It makes no sense.  I guess this is why people use global params for displacements.  There is nothing you can do about this but I find it annoying.

File: modules/combined/test/tests/optimization/invOpt_mechanics/forward_and_adjoint.i [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+[Materials]
+  [elasticity]
+    type = ComputeIsotropicElasticityTensor
+    youngs_modulus = 10e3
+    poissons_ratio = 0.3
+  []
+  # [strain]
+  #   type = ComputeSmallStrain
+  #   displacements = 'disp_x disp_y'
```
Comment (lynnmunday) [inline modules/combined/test/tests/optimization/invOpt_mechanics/forward_and_adjoint.i:88]: remove

File: modules/combined/test/tests/optimization/invOpt_mechanics/forward_and_adjoint.i [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+    type = PointValueSampler
+    variable = 'disp_x disp_y'
+    points = '5.0 1.0 0'
+    sort_by = x
+  []
+[]
+
+[Reporters]
+  [measure_data_x]
+    type = OptimizationData
```
Comment (lynnmunday) [inline modules/combined/test/tests/optimization/invOpt_mechanics/forward_and_adjoint.i:116]: NO weights!  this is awesome.

File: modules/combined/test/tests/optimization/invOpt_mechanics/forward_and_adjoint.i [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+  skip_nl_system_check = true
+[]
+
+[Variables]
+  [adjoint_x]
+    initial_condition = 0
+    solver_sys = adjoint
+    outputs = none
+  []
+
```
Comment (lynnmunday) [inline modules/combined/test/tests/optimization/invOpt_mechanics/forward_and_adjoint.i:177]: remove space

File: modules/combined/test/tests/optimization/invOpt_mechanics/forward_and_adjoint.i [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    solver_sys = adjoint
+    outputs = none
+  []
+[]
+
+[DiracKernels]
+  [pt_x]
+    type = ReporterPointSource
+    variable = adjoint_x
+    x_coord_name = measure_data_x/measurement_xcoord
```
Comment (lynnmunday) [inline modules/combined/test/tests/optimization/invOpt_mechanics/forward_and_adjoint.i:189]: This is so much cleaner than using weighted values

**Review by lynnmunday — COMMENTED** [review]
PLease add a markdown as an example to the optimization module.  This is really good and clean but it takes some explanation to understand.  Look at this one that links an example markdown in optimization with a combined module test:

https://mooseframework.inl.gov/moose/modules/optimization/examples/topology_optimization/3d_mbb.html

File: modules/combined/test/tests/optimization/invOpt_mechanics/forward_and_adjoint.i (hunk 1 above)
Comment (maxnezdyur) [reply modules/combined/test/tests/optimization/invOpt_mechanics/forward_and_adjoint.i:23]: No idea, but agree super annoying.

File: modules/combined/test/tests/optimization/invOpt_mechanics/tests [hunk 6]
```diff
@@ -11,5 +11,17 @@
     max_threads = 1
     # steady solve
     recover = false
+    detail = 'using a Hessian based optimimization,'
+  []
+  [both_disp]
+    type = CSVDiff
+    rel_err = 1e-6
+    abs_zero = 1e-8
```
Comment (lynnmunday) [inline modules/combined/test/tests/optimization/invOpt_mechanics/tests:19]: This is still failing on one of the tests.  I think with these Tao settings:
```  
petsc_options_iname = '-tao_gttol -tao_max_it'
petsc_options_value = '1e-7 500 `
```
You will never be guaranteed to have a gradient below 1e-7.  You should set abs_zero to 1e-6.

**Review by lynnmunday — APPROVED** [review]
Thanks for the documentation Max,  Great job, I approve.  I only have the one comment about fixing the failing test.

File: modules/combined/test/tests/optimization/invOpt_mechanics/tests (hunk 6 above)
Comment (lynnmunday) [reply modules/combined/test/tests/optimization/invOpt_mechanics/tests:19]: We probably have a lot of tests that are using a large value for abs_zero, like 1, to ignore the gradient.  We should be more careful about that and set abs_zero an order of magnitude larger than the tao solver tolerance to make sure it converged.

**Comment by maxnezdyur** [comment]
@dschwen Can you look at this when you get a chance.

**Review by GiudGiud — APPROVED** [review]

**Comment by GiudGiud** [comment]
please squash those last two commits together and then I ll merge unless Daniel has comments. I agree with Lynn that this looks great

**Review by GiudGiud — APPROVED** [review]

**Pull request merged**